File: 98. Validate Binary Search Tree/main.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

import numpy as np
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValidBSTHelper(root, -np.inf, np.inf)
    
    def isValidBSTHelper(self, root: Optional[TreeNode], min: float, max:float) -> bool:
        if root is None:
            return True
        if root.val <= min:
            return False
        if root.val >= max:
            return False
        return self.isValidBSTHelper(root.left, min, root.val) and self.isValidBSTHelper(root.right, root.val, max)


s = Solution()
# Test Case 1:
tree1 = TreeNode(2, TreeNode(1), TreeNode(3))
# Expected: True
print(s.isValidBST(tree1))

# Test Case 2:
tree2 = TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6)))
# Expected: False
print(s.isValidBST(tree2))

File: 98. Validate Binary Search Tree/main.js
const isValidBST = function (root) {
    function dfs(root, min, max) {
        if (!root) return true;

        if (root.val <= min || root.val >= max) return false;

        return dfs(root.left, min, root.val) && dfs(root.right, root.val, max);
    }
    return dfs(root, -Infinity, Infinity);
};


/* Test Case 1: Simple Binary Tree
    *     2
    *    / \
    *  1   3
    *  
*/
const tree1 = {
    val: 2,
    left: {
        val: 1,
        left: null,
        right: null
    },
    right: {
        val: 3,
        left: null,
        right: {
            val: 5,
            left: null,
            right: null
        }
    }
};

console.log(isValidBST(tree1));

/* Test Case 2: Binary Tree with duplicate values
    *     2
    *    / \    
    *   2   3 
    * 
    * 
*/

const tree2 = {
    val: 2,
    left: {
        val: 2,
        left: null,
        right: null
    },
    right: {
        val: 3,
        left: null,
        right: null
    }
};

console.log(isValidBST(tree2));


